plugins/meta/iplink/main.go

```diff
@@ -0,0 +1,223 @@
+	}
+	return netlink.LinkSetUp(link)
+}
+
+func updateResultsMacAddr(config PluginConf, ifName string, newMacAddr string) error {
+	for _, i := range config.PrevResult.Interfaces {
+		if i.Name == ifName {
+			i.Mac = newMacAddr
+		}
+	}
+	return nil
+}
+
+func changePromisc(ifName string, val bool) error {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("Failed to get %s: %v", ifName, err)
+	}
+
+	if val {
+		return netlink.SetPromiscOn(link)
+	}
+	return netlink.SetPromiscOff(link)
+}
+
+func changeMtu(ifName string, mtu int) error {
+	link, err := netlink.LinkByName(ifName)
+	if err != nil {
+		return fmt.Errorf("Failed to get %s: %v", ifName, err)
+	}
+	return netlink.LinkSetMTU(link, mtu)
+}
+
+func parseConf(data []byte, envArgs string) (*PluginConf, error) {
+	conf := PluginConf{IpLinkConf: &IpLinkConf{Promisc: false, Mtu: -1}}
+	if err := json.Unmarshal(data, &conf); err != nil {
+		return nil, fmt.Errorf("failed to load netconf: %v", err)
+	}
+
+	// Parse custom MAC from both env args
+	if envArgs != "" {
+		e := MACEnvArgs{}
+		err := types.LoadArgs(envArgs, &e)
+		if err != nil {
+			return nil, err
+		}
+
+		if e.MAC != "" {
+			conf.Mac = string(e.MAC)
+		}
+	}
+
+	if conf.IpLinkConf == nil && conf.RuntimeConfig.IpLinkConf != nil {
+		conf.IpLinkConf = conf.RuntimeConfig.IpLinkConf
+	}
+
+	// Parse previous result.
+	if conf.RawPrevResult != nil {
+		resultBytes, err := json.Marshal(conf.RawPrevResult)
+		if err != nil {
+			return nil, fmt.Errorf("could not serialize prevResult: %v", err)
+		}
+		res, err := version.NewResult(conf.CNIVersion, resultBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse prevResult: %v", err)
+		}
+		conf.RawPrevResult = nil
+		conf.PrevResult, err = current.NewResultFromResult(res)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert result to current version: %v", err)
+		}
+	}
+
+	return &conf, nil
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConf(args.StdinData, args.Args)
+	if err != nil {
+		return err
+	}
+
+	err = ns.WithNetNSPath(args.Netns, func(_ ns.NetNS) error {
+		var err error
+
+		if conf.Mac != "" {
+			if err = changeMacAddr(args.IfName, conf.Mac); err == nil {
+				err = updateResultsMacAddr(*conf, args.IfName, conf.Mac)
+			}
+		}
+
+		if conf.Promisc != false {
+			if err = changePromisc(args.IfName, true); err != nil {
+				return err
+			}
+		}
+
+		if conf.Mtu != -1 {
+			if err = changeMtu(args.IfName, conf.Mtu); err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+}
+
+func cmdDel(args *skel.CmdArgs) error {
+	// TODO: the settings are not reverted to the previous values. Reverting the
+	// settings is not useful when the whole container goes away but it could be
+	// useful in scenarios where plugins are added and removed at runtime.
```

Comment:
Different plugins can have a different expectations about found situation. Some of them can possibly scream about an error if they will see in netns other configuration than one which they leaved at their end. I don't know any of such plugins which would break because of that at the moment e.g. when mtu is different from configured by it - but there is such possibility, contract is a bit violated.

Reply:
It should be depend on each case, I guess and CNI spec does not take care about it for now.
Currently CNI only provide a way to combination (i.e. main plugin/ipam/meta) and no strict rules to combine plugins. So user can configure flannel + host-local or bridge(with mtu param) and there is no violation from CNI spec point of view.

I believe it should be discussed in CNI community if you' would like to clarify, not this PR.